Mathlib/RingTheory/Ideal/Maps.lean

```diff
@@ -910,3 +910,69 @@ lemma coe_ideal_map (I : Ideal A) :
     Ideal.map f I = Ideal.map (f : A →+* B) I := rfl
 
 end AlgHom
+
+noncomputable section IsPrincipal
+
+namespace Ideal
+
+open nonZeroDivisors Submodule
+
+variable (R : Type*) [CommRing R] [IsDomain R]
+
+/-- The equivalence between `Associates R` and the principal ideals of `R` defined by sending the
+class of `x` to the principal ideal generated by `x`. -/
+noncomputable def associatesEquivIsPrincipal :
```

Comment:
Can this be upgraded to a mul equiv (or something like that, it also presevers `0`)?

Reply:
I could add the instance 
```lean
instance : MonoidWithZero {I : Ideal R // IsPrincipal I}
```
and then proving that it is `MulEquiv` is easy. I don't think there is a `MulEquivWithZero` but adding a separate lemma that states that is equal to zero iff zero should easy too.

Reply:
Maybe a sensible thing to do is to define the submonoid of principal ideals. I don't think it is a lot of work.

Reply:
Sure. I have a question though [Ideal.span_singleton_mul_span_singleton](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Ideal/Operations.html#Ideal.span_singleton_mul_span_singleton) is defined in `Ideal.Operations` so there are several possibilities: 

1. Define the submonoid of principal ideals in this file, but then `associatesEquivIsPrincipal` has to be moved there too and @faenuccio maybe not like that :wink:
2. Move `Ideal.span_singleton_mul_span_singleton` to `Ideal.Maps` but it feel weird to separate it from a bunch of related results
3. Create new file `Ideal.IsPrincipal` and put everything related to principal ideals there.
4. ...

Reply:
I guess creating a new file is a good idea, and quite cheap.

Reply:
My plan is to create a new file`Ideal.IsPrincipal` and move everything (basic) related to `Ideal.span_singleton` and `Ideal.IsPrincipal` (aka `Submodule.IsPrincipal`) in this file. Before I do it though, I want to make sure we are in agreement about this. 
The other option would be to only put in this new file things about the submonoid of principal ideals, but then that file would need to imports all other basic `Ideal` files.

Reply:
I am certainly all in favour of creating `Ideal.IsPrincipal` because many results are quite specific to principal ideals and they deserve a separate file. Especially if this will move these results out of `Ideal.Map` :wink:

Reply:
Well, I'll go ahead and see what it looks like first.

Reply:
Hum, things are getting a lot more complicated than I thought and I do not think that moving everything about `span_singleton` in a `IsPrincipal` file is going to work.  I am going to explore a less ambitious option.

Reply:
I opened #13177 which is provides an easier solution